include/dgl/heterograph_interface.h

```diff
@@ -0,0 +1,346 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file dgl/heterograph_interface.h
+ * \brief DGL heterogeneous graph index class.
+ */
+
+#ifndef DGL_HETEROGRAPH_INTERFACE_H_
+#define DGL_HETEROGRAPH_INTERFACE_H_
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+#include "array.h"
+#include "graph_interface.h"
+
+namespace dgl {
+
+struct HeteroSubgraph;
+
+class HeteroGraphInterface;
+typedef std::shared_ptr<HeteroGraphInterface> HeteroGraphPtr;
+
+class HeteroGraphInterface : public GraphInterface {
+ public:
+  virtual ~HeteroGraphInterface() = default;
+
+  /*!
+   * \brief Add vertices to the graph with the given node type.
+   *
+   * \param ntype the node type
+   * \param num_vertices the number of vertices to add
+   */
+  virtual void AddVertices(dgl_type_t vtype, uint64_t num_vertices);
+
+  /*!
+   * \brief Add one edge to the graph with given edge type.
+   * \param etype the edge type
+   * \param src the type-specific ID of source node
+   * \param dst the type-specific ID of destination node
+   * \note Source node type and destination node type are found according to the
+   * metagraph.
+   */
+  virtual void AddEdge(dgl_type_t etype, dgl_id_t src, dgl_id_t dst) = 0;
+
+  /*!
+   * \brief Add edges with the same edge type to the graph.
+   * \param etype the edge type
+   * \param src_ids the type-specific IDs of source nodes
+   * \param dst_ids the type-specific IDs of destination nodes
+   * \note Source node type and destination node type are found according to the
+   * metagraph.
+   */
+  virtual void AddEdges(dgl_type_t etype, IdArray src_ids, IdArray dst_ids);
+
+  /*! \return the number of vertices of given type in the graph */
+  virtual uint64_t NumVertices(dgl_type_t vtype) const = 0;
+
+  /*! \return the number of edges of given type in the graph */
+  virtual uint64_t NumEdges(dgl_type_t etype) const = 0;
+
+  /*!
+   * \return true if the vertex exists
+   */
+  virtual bool HasVertex(dgl_type_t vtype, dgl_id_t vid) const {
+    return vid < NumVertices(vtype);
+  };
+
+  /*!
+   * \return a boolean array of the existence of the nodes
+   */
+  virtual BoolArray HasVertices(dgl_type_t vtype, IdArray vids) const {
+    // FIXME: duplicate code with GraphInterface::HasVertices due to
+    // additional argument vtype
```

Comment:
There are many similar APIs in heterograph with an extra argument `vtype` or `etype`. For `etype`, could we expose one interface like `GetRelationGraph(etype)` which returns a bipartite graph view, and perform query on the returned graph? This might be able to save a lot of (potentially redundant APIs). For `vtype`, I guess it's fine.